src/main/java/in/projecteka/consentmanager/user/UserService.java

```diff
@@ -229,4 +237,36 @@ public String getUserIdSuffix() {
     public int getExpiryInMinutes() {
         return otpServiceProperties.getExpiryInMinutes();
     }
+
+    public Mono<RecoverCmIdResponse> recoverCmId(RecoverCmIdRequest request) {
+        return isInvalidRecoveryRequest(request)
+                .switchIfEmpty(Mono.defer(() -> Mono.error(ClientError.invalidRecoveryRequest())))
+                .flatMap(validRequest ->
+                        userRepository.getCmIdBy(validRequest.getGender(),
+                                IdentifierUtils.getIdentifierValue(
+                                        validRequest.getVerifiedIdentifiers(),
+                                        IdentifierType.MOBILE)))
+                .flatMap(users -> new NameFilter().filter(users, request.getName()))
+                .flatMap(users -> new YOBFilter().filter(users, request.getYearOfBirth()))
+                .flatMap(users -> new ABPMJAYIdFilter().filter(users, request.getUnverifiedIdentifiers()))
+                .switchIfEmpty(Mono.defer(() -> Mono.error(ClientError.noPatientFound())))
+                .flatMap(this::validateNonZeroRows)
+                .flatMap(user -> Mono.just(RecoverCmIdResponse.builder().cmId(user.getIdentifier()).build()))
+                .switchIfEmpty(Mono.defer(() -> Mono.error(ClientError.multiplePatientsFound())));
+    }
+
+    private boolean isInvalidIdentifierMapped(List<Identifier> identifiers, IdentifierGroup identifierGroup) {
+        return identifiers.stream().anyMatch(identifier -> !identifier.getType().getIdentifierGroup().equals(identifierGroup) || !identifier.getType().isValid(identifier.getValue()));
+    }
+
+    private Mono<RecoverCmIdRequest> isInvalidRecoveryRequest(RecoverCmIdRequest request) { //breakdown
+        boolean areMandatoryFieldsNull = request.getName() == null || request.getGender() == null || !IdentifierUtils.isIdentifierTypePresent(request.getVerifiedIdentifiers(), IdentifierType.MOBILE);
+        boolean isInvalidVerifiedIdentifierMapped = isInvalidIdentifierMapped(request.getVerifiedIdentifiers(), IdentifierGroup.VERIFIED_IDENTIFIER);
+        boolean isInvalidUnverifiedIdentifierMapped = isInvalidIdentifierMapped(request.getUnverifiedIdentifiers(), IdentifierGroup.UNVERIFIED_IDENTIFIER);
+        return areMandatoryFieldsNull || isInvalidVerifiedIdentifierMapped || isInvalidUnverifiedIdentifierMapped ? Mono.empty() : Mono.just(request);
+    }
+
+    private Mono<User> validateNonZeroRows(List<User> rows) {
```

Comment:
if you are returning a Mono<User>, maybe consider renaming the method - getDistinctUser( .. )

Reply:
renamed